hazelcast-hibernate/hazelcast-hibernate3/src/test/java/com/hazelcast/hibernate/RegionFactoryDefaultTest.java

```diff
@@ -46,6 +54,202 @@ protected Properties getCacheProperties() {
+        //hit entitities: all entitties except the deleted one along with their children
+        assertEquals((entityCount-1) * (2 + childCount), sf2.getStatistics().getSecondLevelCacheHitCount());
+    }
+
+
+    @Test
+    public void testUpdateQueryCausesInvalidationOfEntireRegion() {
+        int entityCount = 10;
+        insertDummyEntities(entityCount);
+        Session session = null;
+
+        int res = executeUpdateQueryInTransaction(sf, "UPDATE DummyEntity set name = 'manually-updated' where id=2");
+
+        assertEquals(1, res);
+
+        sf.getStatistics().clear();
+
+        session = sf.openSession();
+        ArrayList<DummyEntity> entities = new ArrayList<DummyEntity>(entityCount);
+        for (int i=0; i<entityCount; i++) {
+            entities.add((DummyEntity)session.get(DummyEntity.class, (long)i));
+        }
+        assertEquals(entityCount * 2, sf.getStatistics().getSecondLevelCacheMissCount());
+        assertEquals(0, sf.getStatistics().getSecondLevelCacheHitCount());
+    }
+
+    @Test
+    public void testUpdateQueryOnNonStrictRWCausesInvalidationOfEntireRegion() {
+        int entityCount = 10;
+        insertDummyNonStrictRWEntities(entityCount, 0);
+        Session session = null;
+
+        int res = executeUpdateQueryInTransaction(sf, "UPDATE DummyEntityNonStrictRW set name = 'manually-updated' where id=2");
+        assertEquals(1, res);
+        sf.getStatistics().clear();
+
+        session = sf.openSession();
+        ArrayList<DummyEntityNonStrictRW> entities = new ArrayList<DummyEntityNonStrictRW>(entityCount);
+        for (int i=0; i<entityCount; i++) {
+            entities.add((DummyEntityNonStrictRW)session.get(DummyEntityNonStrictRW.class, (long) i));
+        }
+        assertEquals(entityCount * 2, sf.getStatistics().getSecondLevelCacheMissCount());
+        assertEquals(0, sf.getStatistics().getSecondLevelCacheHitCount());
+    }
+
+    @Test
+    public void testQueryRegionEviction() {
+        sf.getCache().evictDefaultQueryRegion();
+        int entityCount = 10;
+        insertDummyEntities(entityCount, 0);
+        List<DummyEntity> retrievedEntities = executeListQueryInTransaction(sf);
+        assertEquals(entityCount, retrievedEntities.size());
+
+        sf.getStatistics().clear();
+        executeListQueryInTransaction(sf);
+
+        assertEquals(1, sf.getStatistics().getQueryCacheHitCount());
+        assertEquals(0, sf.getStatistics().getQueryCacheMissCount());
+
+        executeUpdateQueryInTransaction(sf, "delete from DummyEntity");
+
+        sf.getStatistics().clear();
+        executeListQueryInTransaction(sf);
+        executeListQueryInTransaction(sf);
+        assertEquals(1, sf.getStatistics().getQueryCacheHitCount());
+        assertEquals(1, sf.getStatistics().getQueryCacheMissCount());
+    }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void testReadOnlyUpdate() throws Exception{
```

Comment:
The problem in this test is that it isn't clear where the UnsupportedOperationException is thrown!

I normally make it my last statement in my test. And sometimes that isn't even good enough, so I explicitly wrap it around the statement that is supposed to fail. 

Also can't all the tx cleanup not be moved into the tear down? Tear down should kill any not committed/aborted transactions. This way your test remains very clean.